streams: Make default notifications settings customizable per-stream.

Ujjawal3 opened the pull request:
This PR introduces a new boolean field `push_notifications_enabled` in Steam model. If this field is set to true push notifications will be enabled by default for new subscribers of steam. Default value of this field is false.



<!-- Describe your pull request here.-->

Fixes: #23873 

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
ScreenShot showing Stream Creation form for owner

![Screenshot from 2023-01-23 22-06-42](https://user-images.githubusercontent.com/99073049/214096385-cdfb700c-1aae-4edd-bc06-de33da74bdab.png)

Tooltip shown when push notifications are not configured on server.
![Screenshot from 2023-03-02 21-16-15](https://user-images.githubusercontent.com/99073049/222685321-7267df84-aa07-4314-852a-6736a4dc800f.png)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-ste

p guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Ujjawal3 commented:
@alya please review this pr.

alya commented:
Thanks! I found one behavior in manual testing that I think needs to be changed:

1. Unsubscribe a user from a stream
2. Turn on the new default notification setting
3. Resubscribe the same user to the stream

Expected: Notifications are on for that user
Actual: Judging by the notification settings menu, notifications remain off

alya commented:
Let's also rename the setting to "Mobile push notifications enabled by default", and give it more vertical space above, so that it's clear that it's not related to the retention setting. It should have the same amount of space above as "Announce stream".

Ujjawal3 commented:
> Thanks! I found one behavior in manual testing that I think needs to be changed:
> 
> 1. Unsubscribe a user from a stream
> 2. Turn on the new default notification setting
> 3. Resubscribe the same user to the stream
> 
> Expected: Notifications are on for that user Actual: Judging by the notification settings menu, notifications remain off


@alya this behaviour was because as per the current behaviour in Zulip if a user is already subscribed to stream some time in past then on again subscribing the value of push notifications will be the value of push notifications at the time of unsubscription in the past. In my PR also I kept this behaviour. So should I change it?

alya commented:
Ah, hm, I see -- could you please ask for feedback on this question in the development community by starting a thread in #design?

Ujjawal3 commented:
@alya I have started [CZO discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/Resubscribing.20a.20stream.20which.20was.20unsubscribed) regarding the same.

alya commented:
Thanks! Let's add the same warning we have near the personal stream notification settings in the case when notifications are not configured on the server:

![Screen Shot 2023-01-30 at 10 51 37 AM](https://user-images.githubusercontent.com/2090066/215568755-298b4aa8-7851-4e65-a5d4-35a30435a13b.png)

I'm actually thinking the icon in both places should be https://fontawesome.com/v4/icon/exclamation-circle, which can be done as a separate commit on this PR.

alya commented:
I think we can have leave the stream setting toggleable even when notifications aren't enabled on the server, but leaving this note here in case other reviewers have feedback on that.

alya commented:
Once you've implemented the change above, please ping @sahil839 for a review.

Ujjawal3 commented:
@sahil839 please review this PR.

Ujjawal3 commented:
@sahil839 please review this PR.

sahil839 commented on zerver/models.py:
```diff
@@ -2530,6 +2530,8 @@ class Stream(models.Model):
     }
     message_retention_days = models.IntegerField(null=True, default=None)
 
+    # False means we inherit user-level default
```
Would be nice to add here that this will be used to set `push_notifications` field when a user is subscribing.

sahil839 commented on zerver/views/streams.py:
```diff
@@ -602,6 +608,7 @@ def add_subscriptions_backend(
+        stream_dict_copy["push_notifications_enabled"] = push_notifications_enabled
         stream_dicts.append(stream_dict_copy)
 
     # Validation of the streams arguments, including enforcement of
```
You would also need to make sure that non-admin users cannot set `push_notifications_enabled` as `True` similar to what we do for non-owners to not allow them setting `message_retention_days` other than `None`.

sahil839 commented on static/styles/subscriptions.css:
```diff
@@ -1021,6 +1021,10 @@ div.settings-radio-input-parent {
     }
+
+    #push-notifications-enabled {
+        margin: 25px auto;
```
I think the margin of 25px is too much. Maybe we should use 20px which we used for `input-group` divs.

sahil839 commented on static/templates/stream_settings/stream_types.hbs:
```diff
@@ -80,3 +80,16 @@
 {{/if}}
+
+{{#if (or is_admin is_stream_edit)}}
+    <div id="push-notifications-enabled">
```
I think we should disable this setting when the server is not configured to send push notifications.

Ujjawal3 replied:
As mentioned in https://github.com/zulip/zulip/pull/23943#issuecomment-1409162400 I have not disabled it. If this should be disabled let me know I will make the changes.

sahil839 commented on static/templates/stream_settings/stream_settings_checkbox.hbs:
```diff
@@ -21,7 +21,7 @@
     {{/if}}
     {{#if (eq setting_name "push_notifications")}}
-    <i class="fa fa-question-circle settings-info-icon {{#unless disabled_realm_setting}}hide{{/unless}} tippy-zulip-tooltip"
+    <i class="fa fa-exclamation-circle settings-info-icon {{#unless disabled_realm_setting}}hide{{/unless}} tippy-zulip-tooltip"
```
Is this change discussed somewhere? We still use `fa-info-circle` in settings. And we probably would have to instead change the tooltip to show when hovering over the input or lable instead of having a icon like we did recently for `Deactivate organization` button. Till then we can probably use `fa-info-circle` for consistency.

Ujjawal3 replied:
This was discussed in https://github.com/zulip/zulip/pull/23943#issuecomment-1409161116

sahil839 commented on static/templates/stream_settings/stream_types.hbs:
```diff
@@ -80,3 +80,16 @@
+            <span></span>
+            {{t "Mobile push notifications enabled by default" }}
+        </label>
+        <i class="fa fa-exclamation-circle settings-info-icon {{#if realm_push_notifications_enabled}}hide{{/if}} tippy-zulip-tooltip"
```
We can instead show this tooltip while hovering on the input and label as we did for `Deactivate organization` button as we would probably disable this setting if server is not configured to send push notifications. Or we can atleast use `fa-info-circle` so that we can keep it consistent with other settings pages. Though some of the places still use `fa-question-circle`.

Ujjawal3 replied:
I have done the changes. Now tooltip is shown on hovering the whole input and label.

sahil839 commented on static/js/stream_ui_updates.js:
```diff
@@ -119,6 +119,11 @@ export function update_regular_sub_settings(sub) {
+        if (sub.push_notifications) {
+            $("#sub_push_notifications_setting input").prop("checked", true);
+        } else {
+            $("#sub_push_notifications_setting input").prop("checked", false);
```
We should actually remove this and instead call `stream_events.update_property` in `server_events_dispatch.js` or `mark_subscribed` itself to make sure UI in both stream settings and personal notifcations settings get updated.

sahil839 reviewed (commented):
@Ujjawal3 thanks for working on this. Posted a batch of comments.

I am not sure whether we want tooltip on `i` icon or while hovering over the whole setting element (i.e. input and label) when push notifications are not configured on realm or server. @alya thoughts on this?

alya commented:
> I am not sure whether we want tooltip on `i` icon or while hovering over the whole setting element (i.e. input and label) when push notifications are not configured on realm or server. @alya thoughts on this?

The whole setting element makes sense to me.

Ujjawal3 commented on static/js/stream_create.js:
```diff
@@ -240,6 +240,9 @@ function create_stream() {
         $("#announce-new-stream input").prop("checked");
     data.announce = JSON.stringify(announce);
 
+    const push_notifications_enabled = $("#push-notifications-enabled input").is(":checked");
```
Yep ! I have updated the PR. Now `prop("checked")` is used.

Ujjawal3 commented:
@sahil839 I have done the changes as per your comments. Please have a look.

sahil839 commented on web/src/server_events_dispatch.js:
```diff
@@ -532,6 +532,7 @@ export function dispatch_normal_event(event) {
                     for (const rec of event.subscriptions) {
                         const sub = sub_store.get(rec.stream_id);
                         if (sub) {
+                            sub.push_notifications = rec.push_notifications;
```
We should instead set `push_notifcations` in `mark_subscribed`.

sahil839 commented on web/src/stream_ui_updates.js:
```diff
@@ -119,6 +120,7 @@ export function update_regular_sub_settings(sub) {
     if (sub.subscribed) {
         $settings.find(".personal_settings").addClass("in");
         $settings.find(".stream-email-box").show();
+        stream_events.update_property(sub.stream_id, "push_notifications", sub.push_notifications);
```
I think we can also remove this, if we set `push_notifications` inside `mark_susbcribed` just like we set `color` of the stream.

sahil839 reviewed (commented):
Posted some more comments.

sahil839 commented on zerver/tests/test_subs.py:
```diff
@@ -4778,7 +4932,7 @@ def test_bulk_subscribe_many(self) -> None:
         test_user_ids = [user.id for user in test_users]
 
-        with self.assert_database_query_count(20):
+        with self.assert_database_query_count(50):
```
This is due to calling `save` in loop as mentioned [here](https://github.com/zulip/zulip/pull/23943#discussion_r1119802190).

Ujjawal3 commented:
@sahil839 I have done the changes as per your comments.

sahil839 commented on web/src/stream_events.js:
```diff
@@ -114,6 +117,13 @@ export function mark_subscribed(sub, subscribers, color) {
+    if (sub.push_notifications_enabled) {
+        sub.push_notifications = sub.push_notifications_enabled;
+    } else {
+        sub.push_notifications = null;
```
Can we instead just set `push_notifications` field to the `push_notifications` field in the object sent with the event.

sahil839 commented on zerver/actions/streams.py:
```diff
@@ -526,21 +527,34 @@ def bulk_add_subscriptions(
                     subs_to_activate.append(sub_info)
 
+        subs = [info.sub for info in subs_to_activate]
+        Subscription.objects.bulk_update(subs, ["push_notifications"])
```
I think we should call `bulk_update` outside the `for user_profile in users:` loop so that we can just update all the `Subscription` objects by single update to database.

sahil839 commented on web/templates/stream_settings/stream_types.hbs:
```diff
@@ -80,3 +80,15 @@
+            <span></span>
+            {{t "Mobile push notifications enabled by default" }}
+        </label>
+        <i class="fa fa-exclamation-circle settings-info-icon {{#if realm_push_notifications_enabled}}hide{{/if}}">
```
We can remove this icon now since we show the tooltip for whole element.

sahil839 commented on web/templates/stream_settings/stream_types.hbs:
```diff
@@ -80,3 +80,15 @@
 {{/if}}
+
+{{#if (or is_admin is_stream_edit)}}
+    <div id="push-notifications-enabled" {{#unless realm_push_notifications_enabled}}class="show-disabled-tooltip"{{/unless}}>
```
Also, can we use `stream_settings_checkbox` template here?

sahil839 replied:
Or maybe `settings_checkbox` to avoid complexity in save-discard widget code.

sahil839 commented on web/styles/subscriptions.css:
```diff
@@ -1010,6 +1010,11 @@ div.settings-radio-input-parent {
+
+    #push-notifications-enabled {
+        margin: 20px 0;
+        width: fit-content;
```
Is this `width` property required here?

sahil839 commented:
Posted few comments.

> I think we can have leave the stream setting toggleable even when notifications aren't enabled on the server, but leaving this note here in case other reviewers have feedback on that.

I think it would be better if we disable the setting if push notifications are not configured on server. We also disable the setting in `Personal` section.

Ujjawal3 commented:
@sahil839 I have addressed your comments. 
Also now checkbox is disabled if `push_notifications` are not configured. The UI for disabled state is same as we have in personal settings.

Ujjawal3 commented:
@sahil839 bumping this PR for review, as it has been a while since last review.

sahil839 commented on web/src/stream_events.js:
```diff
@@ -113,6 +116,8 @@ export function mark_subscribed(sub, subscribers, color) {
         return;
     }
+    sub.push_notifications = push_notifications;
+    update_property(sub.stream_id, "push_notifications", sub.push_notifications);
```
You can just call the `update_push_notifications_enabled` like we do for the `color` property, and we only call `update_property` on recieving event. Also, I think calling that function directly would be better for readability.

sahil839 replied:
Oh, sorry for the confusion. Ok, let's keep this as it is for now.

sahil839 reviewed (commented):
Posted a few comments. Also, I think you missed to reply or update the PR as per couple of previous comments.

I also noticed a bug where the input is enabled for non-admins in stream edit section.

Ujjawal3 commented:
> Posted a few comments. Also, I think you missed to reply or update the PR as per couple of previous comments.
> 
> I also noticed a bug where the input is enabled for non-admins in stream edit section.

Yes, I updated the PR as per the comments but missed to reply some which I have replied now. Also thanks for reporting that bug. I have fixed it now.

sahil839 commented on zerver/actions/streams.py:
```diff
@@ -1357,3 +1371,40 @@ def do_change_stream_group_based_setting(
+
+@transaction.atomic(durable=True)
+def do_change_push_notifications_enabled(
+    stream: Stream, acting_user: UserProfile, push_notifications_enabled: bool
```
Would be better to keep `acting_user` as mandatory keyword argument like we do for other similar functions.

sahil839 reviewed (commented):
Posted one comment. There are still a couple of comments like - https://github.com/zulip/zulip/pull/23943#discussion_r1134220878, https://github.com/zulip/zulip/pull/23943#discussion_r1134114317 and https://github.com/zulip/zulip/pull/23943#discussion_r1122076292, where there is no change and no reply. Please carefully check whether all the comments are responded (either by comment or update in the PR) before asking for the review.

Also please update the commit messages to follow the [guidelines](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-messages).

Ujjawal3 commented:
> Posted one comment. There are still a couple of comments like - [#23943 (comment)](https://github.com/zulip/zulip/pull/23943#discussion_r1134220878), [#23943 (comment)](https://github.com/zulip/zulip/pull/23943#discussion_r1134114317) and [#23943 (comment)](https://github.com/zulip/zulip/pull/23943#discussion_r1122076292), where there is no change and no reply. Please carefully check whether all the comments are responded (either by comment or update in the PR) before asking for the review.
> 
> Also please update the commit messages to follow the [guidelines](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-messages).

I replied to all these comments. But I don't know why there was a pending label on the replies and the replies were not visible. I have fixed it know and replies are visible.
My apologies I was not aware that pending label on reply means the reply is not posted yet.

sahil839 commented on zerver/actions/streams.py:
```diff
@@ -526,6 +527,10 @@ def bulk_add_subscriptions(
+                    if not stream.push_notifications_enabled:
+                        sub_info.sub.push_notifications = None
+                    else:
+                        sub_info.sub.push_notifications = True
```
I am not sure whether we also want `RealmAuditLog` objects for change in `push_notifications`. Let us not make any changes now and wait for feedback from Tim.

sahil839 commented on zerver/tests/test_subs.py:
```diff
@@ -4778,7 +4932,7 @@ def test_bulk_subscribe_many(self) -> None:
         test_user_ids = [user.id for user in test_users]
 
-        with self.assert_database_query_count(20):
+        with self.assert_database_query_count(50):
```
Noting here that this was fixed by using `bulk_update`.

sahil839 commented:
Ok, looks good to me overall. Just posted one comment about test code which I am not sure about the need to check that case.

@alya do you want to have a look again? Just a note that, this PR has been updated to disable the setting when push notifications are not configured on the server for now, which I think is better since we also disable the setting in `Personal` section, for which this setting acts as a default.

sahil839 commented:
> I replied to all these comments. But I don't know why there was a pending label on the replies and the replies were not visible.

You might have clicked on `Start a review` button instead of `Add single comment` when posting comments. In that case, the comments posted are all pending unless you submit your review.

alya commented:
Looks like you need to rebase the PR and renumber the migrations. I'm seeing this when I provision:

```
Conflicting migrations detected; multiple leaf nodes in the migration graph: (0433_stream_push_notifications_enabled, 0434_create_nobody_system_group in zerver).
To fix them run 'python manage.py makemigrations --merge'

Subcommand of /srv/zulip/tools/lib/provision_inner.py failed with exit status 1: env DJANGO_SETTINGS_MODULE=zproject.settings ZULIP_DB_NAME=zulip ./manage.py migrate --no-input
Actual error output for the subcommand is just above this.
```

alya commented:
> @alya do you want to have a look again? Just a note that, this PR has been updated to disable the setting when push notifications are not configured on the server for now, which I think is better since we also disable the setting in Personal section, for which this setting acts as a default.

I tested this part; makes sense to me.

I wasn't sure how to retest the other behaviors now, but I think we can move on to integration review. @timabbott This PR has been reviewed by @sahil839 as well.

timabbott commented:
Closing in favor of https://github.com/zulip/zulip/issues/23873, which continues this effort. Thanks @Ujjawal3!
